### PR TITLE
Upgrading single chip MNIST training code to use cross entropy loss.

### PR DESCRIPTION
**Pull request by umalesTT**

### CODE REVIEW

File: blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py [hunk 1]
```diff
@@ -213,11 +214,22 @@ def train(config_path=None):
         checkpoint_file_path = os.path.join(checkpoint_dir, checkpoint_file_name)
         save_checkpoint(checkpoint_file_path, state, epoch)
 
+        if eval_batch_metrics_avg["loss"] < best_val_loss - early_stopping_config.min_delta:
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py:217]: the logic here differs a bit from multichip case (there is no fallback to best params)

File: blacksmith/experiments/jax/mnist/single_chip/test_pure_jax_mnist.py [hunk 2]
```diff
@@ -159,10 +167,18 @@ def train_mlp(
...
+                learning_rate_host = jax.device_put(learning_rate, jax.devices("cpu")[0])
 
-                batch_accuracy = compute_accuracy(logits, y_batch)
+                # Optimizer step is done on CPU (https://github.com/tenstorrent/tt-xla/issues/342)
+                params_host_updated = update(params_host, grads_host, learning_rate_host)
+
+                params = jax.device_put(params_host_updated, current_device)
+
+                logits_host = jax.device_put(logits, jax.devices("cpu")[0])
+                # Accuracy calculation is done on CPU, as argmax is not supported on multi-device (https://github.com/tenstorrent/tt-mlir/issues/3963)
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/jax/mnist/single_chip/test_pure_jax_mnist.py:180]: does argmax work on single-device? It should be on device if it works.

File: blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py (hunk 1 above)
Comment (vladimirjovanovicTT) [reply blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py:217]: pure jax version is correct, though?

File: blacksmith/experiments/jax/mnist/single_chip/train_utils/train_functions.py [hunk 3]
```diff
@@ -9,51 +9,41 @@
...
-@jax.jit
-def forward_pass(params, x):
-    return MLP().apply({"params": params}, x, mutable=["params"])
-
-
-def forward_and_compute_loss(params, x, y):
-    logits, _ = forward_pass(params, x)
-    loss = func_optax_loss(logits, y)
-    return loss
+def func_optax_loss(logits, labels):
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/jax/mnist/single_chip/train_utils/train_functions.py:12]: this method should have cross_entropy in its name

**Review by vladimirjovanovicTT — CHANGES_REQUESTED** [review]

File: blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py [hunk 4]
```diff
@@ -150,16 +149,22 @@ def train(config_path=None):
...
 
-            state, loss, grads = train_step(state, batch_images, batch_labels)
-            logits = eval_step(state.params, batch_images)
+            loss, grads, logits = compute_loss_grads_and_logits(state.params, batch_images, batch_labels)
 
+            grads_host = jax.device_put(grads, cpu_device)
             logits_host = jax.device_put(logits, cpu_device)
             batch_labels_host = jax.device_put(batch_labels, cpu_device)
 
             # Accuracy calculation is done on CPU, as argmax is not supported on device (https://github.com/tenstorrent/tt-metal/issues/20570)
```
[the remark this answers is not in the document]
Comment (umalesTT) [reply blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py:158]: Yes

File: blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py (hunk 1 above)
Comment (umalesTT) [reply blacksmith/experiments/jax/mnist/single_chip/test_flax_mnist.py:217]: It's actually the same logic; we do fallback to best params through loading a model state from best epoch

File: blacksmith/experiments/jax/mnist/single_chip/test_pure_jax_mnist.py (hunk 2 above)
Comment (umalesTT) [reply blacksmith/experiments/jax/mnist/single_chip/test_pure_jax_mnist.py:180]: Yes it is on device now
